### ISSUE TITLE
jacrev uses lots of GPU memory

### ISSUE THREAD

**Issue by cyyever** [pytorch/functorch#636]
I want to get the derivative vector of the gradient of M(X,Y) to X_0, so I write a simple script to verify it, but my implementation runs out of memory. Is there a better way to achieve this goal?

```
import collections

import torch
import torch.nn as nn
import torch.nn.functional as F


def cat_tensors_to_vector(tensors: list) -> torch.Tensor:
    return nn.utils.parameters_to_vector([t.reshape(-1) for t in tensors])


from functorch import grad, jacfwd, jacrev, make_functional


class LeNet5(nn.Module):
    """
    Input - 1x32x32
    C1 - 6@28x28 (5x5 kernel)
    tanh
    S2 - 6@14x14 (2x2 kernel, stride 2) Subsampling
    C3 - 16@10x10 (5x5 kernel, complicated shit)
    tanh
    S4 - 16@5x5 (2x2 kernel, stride 2) Subsampling
    C5 - 120@1x1 (5x5 kernel)
    F6 - 84
    tanh
    F7 - 10 (Output)
    """

    input_size = (32, 32)

    def __init__(self, input_channels=1):
        super().__init__()
        self.input_channels = input_channels
        self.convnet = nn.Sequential(
            collections.OrderedDict(
                [
                    ("c1", nn.Conv2d(self.input_channels, 6, kernel_size=5)),
                    ("relu1", nn.ReLU()),
                    ("s2", nn.MaxPool2d(kernel_size=2, stride=2)),
                    ("c3", nn.Conv2d(6, 16, kernel_size=5)),
                    ("relu3", nn.ReLU()),
                    ("s4", nn.MaxPool2d(kernel_size=2, stride=2)),
                    ("c5", nn.Conv2d(16, 120, kernel_size=5)),
                    ("relu5", nn.ReLU()),
                ]
            )
        )

        self.fc = nn.Sequential(
            collections.OrderedDict(
                [
                    ("f6", nn.Linear(120, 84)),
                    ("relu6", nn.ReLU()),
                    ("f7", nn.Linear(84, 10)),
                ]
            )
        )

    def forward(self, x):
        output = self.convnet(x)
        output = output.view(x.size(0), -1)
        output = self.fc(output)
        return output


model = LeNet5()
model.cuda()

fun, param = make_functional(model)


def fun2(param, input_tensor, target, input_index, elem):
    shape = input_tensor.shape
    new_tensor = cat_tensors_to_vector(
        [
            input_tensor.view(-1)[:input_index],
            elem,
            input_tensor.view(-1)[input_index + 1:],
        ]
    )
    input_tensor = new_tensor.reshape(shape)
    return F.cross_entropy(fun(param, input_tensor), target)


input_tensor = torch.rand((1, 32, 32)).cuda()
target = torch.LongTensor([1]).cuda()

print(
    jacrev(grad(fun2, argnums=0), argnums=4)(
        param, input_tensor, target, 5, input_tensor.view(-1)[5]
    ).sum()
)

```
And it failed with
```
RuntimeError: CUDA out of memory. Tried to allocate 11.04 GiB (GPU 0; 10.76 GiB total capacity; 606.91 MiB already allocated; 9.05 GiB free; 624.00 MiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

**Comment by vfdev-5** [comment]
I could repro the issue and looks like it is combination of grad + jacrev which leads to OOM.
We had a similar situation with `vhp` for while benchmarking on other models.
In particular, `_construct_standard_basis_for` tries to allocate on device 10 2d-tensors of size:
```
(150, 6, 2400, 16, 48000, 120, 10080, 84, 840, 10)
```
which takes around 9GB

**Comment by cyyever** [comment]
@vfdev-5  Thanks your reply, I am not sure what the code does. But in MATH the result should be the same size as the gradient

**Comment by vfdev-5** [comment]
@cyyever there is no problems with your code, I was just analysing where OOM comes from and found that it is coming from constructing standard basic tensors on cuda. 
I'm not sure quite sure right now how to handle the issue. Let's wait for a comment from @zou3519

**Comment by cyyever** [comment]
@zou3519 That is too slow. CrossEntropy is a common output function, is there any difficulty to support  forward-mode AD support? I have to try solve this issue by myself...

**Comment by zou3519** [comment]
My guess is that we are going to need forward-mode AD support for cross_entropy's backward function, max_pool2d's backward function, and conv2d's backward function (we can prioritize these).

Could you please let us know a little more about your use case? Do you just want the gradient of M(X,Y) to X_0, or do you want the gradient of M(X, Y) to X_i for all i? Are you trying to get this to work with other models as well, or just LeNet5?

**Comment by cyyever** [comment]
@zou3519 
I have to use cases, the first is  compute the grad to a input scalar, the second is compute jvp of the grad to some vector. Now the first cast is slow and consumes lot of GPU, and I can't find a way to composite the methods provided by functorch to deal with the second case.

**Comment by zou3519** [comment]
>  I can't find a way to composite the methods provided by functorch to deal with the second case because functorch.vjp lacks a argnums parameter

Could you explain a bit more about what you would want an argnums parameter on functorch.vjp would do? Do you want to differentiate with respect to only some of the inputs of the function, or do you want to pass None gradients?

For the first one (differentiating with respect to only some of the inputs of the function), one can pass vjp a new function that captures inputs that we shouldn't differentiate. For example:

```
import torch
from functorch import vjp

def model(param, x):
  return x @ param

x = torch.randn(3)
param = torch.randn(3, 3)

def f(param):
  return model(param, x)
  
output, vjp_fn = vjp(f, param)
grad_output = torch.randn_like(output)
vjp_fn(grad_output)
```

**Comment by cyyever** [comment]
@zou3519  What I want is to compute vjp where the Jacobian matrix is the partial derivative of the gradient function to the input dimension, like this
```
import torch

from functorch import grad, jacrev, vjp


def model(param, x):
    return x[0] * param[0]


x = torch.randn(2)
param = torch.randn(3)


def f(x):
    return grad(model)(param, x)


output, vjp_fn = vjp(f, x)
res = vjp_fn(torch.randn_like(x))
```

here model is a function that takes a parameter vector and an input vector and returns a scalar. However, it fails with
```
RuntimeError: Mismatch in shape: grad_output[0] has a shape of torch.Size([2]) and output[0] has a shape of torch.Size([3]).
```

**Comment by zou3519** [comment]
Thanks for the detailed repro and explanation, @cyyever. That looks like a bug on our side, we'll look into it more

**Comment by zou3519** [comment]
@cyyever I looked into it a bit more. I think the error message makes sense in the example and that there is no bug here. Diving into the example a bit more,

```
import torch
from functorch import grad, jacrev, vjp

def model(param, x):
    return x[0] * param[0]

x = torch.randn(2)
param = torch.randn(3)

def f(x):
    return grad(model)(param, x)

output, vjp_fn = vjp(f, x)
res = vjp_fn(torch.randn_like(x))
```
f is a function with jacobian of shape 3 x 2. Now, the code calls `vjp` which is computing a vector-Jacobian product. Therefore we would expect the vector to have shape [3].

The Tensor being passed in, `torch.randn_like(x)`, has shape [2], so that is where the error message is coming from.

If the quantity you wish to compute is the jacobian-vector product, then we should be using `functorch.jvp` here (and not functorch.vjp). That being said, we did mention above that PyTorch forward-mode AD hasn't implemented many of the forward-mode AD formulas necessary for the computation you wanted (and we can prioritize these)

**Comment by BichenWuUCB** [comment]
I also encountered this issue. I was trying to compute batched per-sample Jacobian of a function y=f(x), where y has the shape of (B, 27) and x has the shape of (B, 3), B is the batch size. The function f() is essentially a neural network but also contains a function torch.nn.functional.grid_sample(). I used `vmap(jacrev(f))(x)` to compute batched per-sample Jacobian, with the expected shape of (B, 27, 3). However, the batch size B has to be very small (~10) or it will trigger CUDA OOM error. Theoretically, it should not consume such high memories. 

I also tried to replace jacrev to jacfwd, but get the following error message: `NotImplementedError: Trying to use forward AD with cudnn_grid_sampler that does not support it.`

Any suggestions on how to resolve this? Thanks!
